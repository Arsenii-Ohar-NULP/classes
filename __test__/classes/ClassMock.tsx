import Class from 'components/classes/Class';
import React from 'react';
export default function ClassMock({ cls }: { cls: Class }) {
  return <div data-testid={`class-${cls.id}`}>{cls.title}</div>;
}
