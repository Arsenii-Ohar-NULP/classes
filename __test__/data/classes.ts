import Class from 'components/classes/Class';

export const sampleFiveClasses: Class[] = [
    {
        id: 1,
        title: 'Class 1',
        description: 'This is class 1',
        teacher_id: 1,
        students_number: 1
    },
    {
        id: 2,
        title: 'Class 2',
        description: 'This is class 2',
        teacher_id: 2,
        students_number: 1
    },
    {
        id: 3,
        title: 'Class 3',
        description: 'This is class 3',
        teacher_id: 3,
        students_number: 1

    },
    {
        id: 4,
        title: 'Class 4',
        description: 'This is class 4',
        teacher_id: 4,
        students_number: 1
    },
    {
        id: 5,
        title: 'Class 5',
        description: 'This is class 5',
        teacher_id: 5,
        students_number: 1
    }
]