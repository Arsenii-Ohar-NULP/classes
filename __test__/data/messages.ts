import Message from 'pages/class/Message';

const sampleMessages : Message[] = [
    {
        id: 1,
        content: 'First',
        date: 'Yesterday',
        user: 1,
        cls: 1,
        fullname: 'Arsenii Ohar',
        username: 'arseniyohar'
    },
    {
        id: 2,
        content: 'First',
        date: 'Yesterday',
        user: 1,
        cls: 1,
        fullname: 'Arsenii Ohar',
        username: 'arseniyohar'
    },
    {
        id: 3,
        content: 'First',
        date: 'Yesterday',
        user: 1,
        cls: 1,
        fullname: 'Arsenii Ohar',
        username: 'arseniyohar'
    }

]