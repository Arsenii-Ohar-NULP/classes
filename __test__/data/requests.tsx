import { JoinRequest } from 'pages/class/JoinRequest'

export const getRequestsWithClassId = (classId: number, number: number) => {
    const requests: JoinRequest[] = []
    for (let i = 0; i < number; i++){
        requests.push(
            {
                userId: i + 1,
                classId: classId
            }
        )
    }

    return requests;
}