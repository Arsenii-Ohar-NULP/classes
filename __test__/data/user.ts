import User, { Role } from 'components/account/User';

export const sampleUser: User = {
    id: 1,
    username: 'seniorohar',
    firstName: 'Arsenii',
    lastName: 'Ohar',
    email: 'arsen.ogar@gmail.com',
    phone: '+380987669293',
    role: Role.Teacher   
}

export const sampleFiveUsers: User[] = [
    {
        id: 1,
        username: 'seniorohar',
        firstName: 'Arsenii',
        lastName: 'Ohar',
        email: 'arsen.ogar@gmail.com',
        phone: '+380987669293',
        role: Role.Student
    },
    {
        id: 2,
    username: 'vitalikpikh',
    firstName: 'Vitalik',
    lastName: 'Pikhotskiy',
    email: 'vitaliipik@gmail.com',
    phone: '+380967629253',
    role: Role.Student
    },
    {
        id: 3,
    username: 'svyatoslavshainoha',
    firstName: 'Svyatoslav',
    lastName: 'Shainoha',
    email: 'vitaliipik@gmail.com',
    phone: '+380967629253',
    role: Role.Student
    },
      {
        id: 4,
    username: 'olehrysin',
    firstName: 'Oleh',
    lastName: 'Rysin',
    email: 'olehrysin@gmail.com',
    phone: '+380967323273',
    role: Role.Student
    },
    {
        id: 5,
    username: 'johnadams',
    firstName: 'John',
    lastName: 'Adams',
    email: 'johnadams@gmail.com',
    phone: '+380967723273',
    role: Role.Student
    }
]