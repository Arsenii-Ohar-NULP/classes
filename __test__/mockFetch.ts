export const mockFetch = () => {
    global.fetch = jest.fn();
}