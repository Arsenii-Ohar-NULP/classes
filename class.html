<!DOCTYPE html>
<!--[if lt IE 7]>      <html class="no-js lt-ie9 lt-ie8 lt-ie7"> <![endif]-->
<!--[if IE 7]>         <html class="no-js lt-ie9 lt-ie8"> <![endif]-->
<!--[if IE 8]>         <html class="no-js lt-ie9"> <![endif]-->
<!--[if gt IE 8]>      <html class="no-js"> <!--<![endif]-->
<html>
  <head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <title></title>
    <meta name="description" content="" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <link rel="stylesheet" href="class.css" />
  </head>
  <body>
    <!--[if lt IE 7]>
      <p class="browsehappy">
        You are using an <strong>outdated</strong> browser. Please
        <a href="#">upgrade your browser</a> to improve your experience.
      </p>
    <![endif]-->
    <nav class="navbar navbar-expand-lg bg-body-tertiary" data-bs-theme="dark">
      <div class="container-fluid">
        <a class="navbar-brand" href="#">Classes</a>
        <button
          class="navbar-toggler"
          type="button"
          data-bs-toggle="collapse"
          data-bs-target="#navbarNavDropdown"
          aria-controls="navbarNavDropdown"
          aria-expanded="false"
          aria-label="Toggle navigation"
        >
          <span class="navbar-toggler-icon"></span>
        </button>
        <div
          class="collapse navbar-collapse justify-content-between"
          id="navbarNavDropdown"
        >
          <ul class="navbar-nav">
            <li class="nav-item">
              <a class="nav-link active" aria-current="page" href="#">Home</a>
            </li>
          </ul>
          <ul class="navbar-nav">
            <li class="nav-item">
              <a class="nav-link" href="#">
                <img
                  src="images/how-to-watch-john-wick (2).jpg"
                  height="48"
                  width="48"
                  class="rounded-circle"
                />
              </a>
            </li>
          </ul>
        </div>
      </div>
    </nav>
    <div class="container-sw">
      <div class="container bg-dark text-white w-auto rounded-1 pb-2">
        <div class="hstack justify-content-between">
          <div>
            <div class="p-3 pb-0">
              <p class="fs-3">Learn React.js in a month</p>
            </div>
            <div class="px-3">
              <a>
                <p class="fs-5">Arsenii Ohar</p>
              </a>
            </div>
            <div class="px-3">
              <p class="fs-6">You can learn React.js here quickly.</p>
            </div>
          </div>
          <button class="btn btn-primary mx-3">Join</button>
        </div>
      </div>
      <div class="bg-body-tertiary">
        <div class="container my-2">
          <div class="hstack gap-3">
            <img
              src="./images/the-rock-small.jpg"
              class="rounded-circle"
              width="64"
              height="64"
            />
            <div class="card bg-secondary h-100 p-2 rounded-5">
              <div class="text-white p-2">
                <div class="vstack mx-3 justify-content-start">
                  <p class="fs-5 mb-0">
                    The Rock <span class="badge badge-dark">Student</span>
                  </p>
                  <p class="fs-6 mb-0">
                    Hey everyone, it's The Rock, and I'm thrilled to be taking
                    this online course on React.js, because I have a specific
                    goal in mind. I want to build an amazing online course
                    landing website, and I know that React.js is the perfect
                    technology for the job.
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>

        <div class="container my-2">
          <div class="hstack gap-3">
            <img
              src="./images/the-rock-small.jpg"
              class="rounded-circle"
              width="64"
              height="64"
            />
            <div class="card bg-secondary h-100 p-2 rounded-5">
              <div class="text-white p-2">
                <div class="vstack mx-3 justify-content-start">
                  <p class="fs-5 mb-0">
                    The Rock <span class="badge badge-dark">Student</span>
                  </p>
                  <p class="fs-6 mb-0">
                    I'm excited to learn all about React.js, from the
                    fundamentals to the more advanced topics. As The Rock, I
                    know that with the skills and knowledge I'll gain from this
                    course, I'll be able to create an incredible user experience
                    for my online course landing website.
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>

        <div class="container my-2">
          <div class="hstack gap-3">
            <img
              src="./images/the-rock-small.jpg"
              class="rounded-circle"
              width="64"
              height="64"
            />
            <div class="card bg-secondary h-100 p-2 rounded-5">
              <div class="text-white p-2">
                <div class="vstack mx-3 justify-content-start">
                  <p class="fs-5 mb-0">
                    The Rock <span class="badge badge-dark">Student</span>
                  </p>
                  <p class="fs-6 mb-0">
                    As The Rock, I know the importance of creating a strong
                    online presence, and I'm confident that this course will
                    help me achieve my goals. With React.js, I'll be able to
                    create a website that not only looks great, but functions
                    flawlessly too.
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>

        <div class="container my-2">
          <div class="hstack gap-3">
            <img
              src="./images/the-rock-small.jpg"
              class="rounded-circle"
              width="64"
              height="64"
            />
            <div class="card bg-secondary h-100 p-2 rounded-5">
              <div class="text-white p-2">
                <div class="vstack mx-3 justify-content-start">
                  <p class="fs-5 mb-0">
                    The Rock <span class="badge badge-dark">Student</span>
                  </p>
                  <p class="fs-6 mb-0">
                    I'm grateful for the opportunity to learn from the best in
                    the business, and as The Rock, I can't wait to put my new
                    skills to the test. With the knowledge I'll gain from this
                    course, I know that I'll be able to create an online course
                    landing website that truly stands out from the crowd.
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>

    <script src="" async defer></script>
  </body>
</html>
