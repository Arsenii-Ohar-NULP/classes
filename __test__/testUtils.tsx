import React, { PropsWithChildren } from 'react';
import { render, renderHook } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import type { PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import type { RootState, AppStore } from 'pages/redux/store';
// As a basic setup, import your same slice reducers
import { authInitialState, authSlice } from 'pages/redux/auth';
import { classSlice, classesInitialState } from 'pages/redux/classes';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = { auth: authInitialState, classes: classesInitialState },
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { auth: authSlice.reducer, classes: classSlice.reducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderHookWithProviders(
  ui,
  {
    preloadedState = { auth: authInitialState, classes: classesInitialState },
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { auth: authSlice.reducer, classes: classSlice.reducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...renderHook(() => ui(), { wrapper: Wrapper, ...renderOptions }) };
}
