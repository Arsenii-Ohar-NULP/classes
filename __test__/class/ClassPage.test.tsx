import React from 'react';
import {sampleFiveClasses} from '__test__/data/classes';
import {sampleUser} from '__test__/data/user';
import {renderWithProviders} from '__test__/testUtils';
import {findClass} from 'components/class/ClassService';
import ClassPage from 'app/main/classes/[id]/page';
import {AuthStatus} from 'components/redux/auth';
import {fireEvent, screen, waitFor} from '@testing-library/react';
import {useRouter} from 'next/navigation';

jest.mock('components/classes/ClassThumbnail');
jest.mock('components/utils/hooks');
jest.mock('components/class/MessagesBar');
jest.mock('components/class/DeleteClassButton');
jest.mock('components/class/ClassService');
jest.mock('components/utils/socket')

const pushMock = jest.fn((url) => console.log(url));
jest.mock('next/navigation', () => ({
    useRouter: () => ({
        pathname: '/',
        push: pushMock,
        replace: pushMock,
        query: {id: '1'},
    }),
}));

describe('class page test', () => {
    it('should match a snapshot and show class info', async () => {
        const cls = sampleFiveClasses[0];
        jest.mocked(findClass).mockResolvedValueOnce(cls);
        const page = renderWithProviders(<ClassPage params={{id: '1'}}/>, {
            preloadedState: {
                auth: {status: AuthStatus.LOGGED_IN, user: sampleUser},
                classes: {
                    userClasses: sampleFiveClasses,
                    joinRequests: [{classId: cls.id, userId: 2}],
                },
                search: {classes: null, students: null},
            },
        });
        await waitFor(() => {
            expect(page.container).toMatchSnapshot();
            screen.findByText(cls.description);
            screen.findByText(cls.title);
        });
    });

    it('should navigate to /requests, when a user is a teacher and requests button has been clicked', async () => {
        const cls = sampleFiveClasses[0];
        jest.mocked(findClass).mockResolvedValueOnce(cls);
        renderWithProviders(<ClassPage params={{id: '1'}}/>, {
            preloadedState: {
                auth: {status: AuthStatus.LOGGED_IN_FETCHED, user: sampleUser},
                classes: {userClasses: sampleFiveClasses, joinRequests: []},
                search: {classes: null, students: null},
            },
        });
        await waitFor(() => {
            const requestsButton = screen.getByRole('button', {
                name: 'Requests',
            });
            fireEvent.click(requestsButton);
        })

        await waitFor(() => {
            expect(useRouter().replace).toHaveBeenCalledWith(`/main/classes/requests/${cls.id}`);
        });
    });
});
