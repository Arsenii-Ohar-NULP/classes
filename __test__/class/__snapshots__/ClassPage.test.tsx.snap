// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class page test should match a snapshot and show class info 1`] = `
<div>
  <div
    class="h-100 d-flex justify-content-center align-items-center"
  >
    <div
      class=""
    >
      <div
        class="spinner-border text-primary text-center"
        role="status"
      >
        <span
          class="visually-hidden"
        >
          Loading...
        </span>
      </div>
    </div>
  </div>
</div>
`;
